[IRGen] Other instructions whose ops/results have packs may allocate pack metadata.

nate-chandler opened the pull request:
Based on https://github.com/apple/swift/pull/67567 .

Instead of assuming that the list of instructions known to allocate pack metadata is exhaustive and returning false from mayRequirePackMetadata for all others, consider the types of the results and operands of other instructions and look for packs.

nate-chandler committed:
[IRGen] alloc_stack may allocate pack metadata.

When it's a tuple containing a pack.

nate-chandler committed:
[AST] PackArchetypes have concrete packs.

Add the recursive property.

nate-chandler committed:
[IRGen] Insts with packs in ops/results may alloc.

Instead of assuming that the list of instructions known to allocate pack
metadata is exhaustive and returning false from mayRequirePackMetadata
for all others, consider the types of the results and operands of other
instructions and look for packs.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test linux platform

slavapestov commented on lib/AST/Type.cpp:
```diff
@@ -3703,8 +3703,9 @@ PackArchetypeType::PackArchetypeType(
-                    RecursiveTypeProperties::HasArchetype, InterfaceType,
-                    ConformsTo, Superclass, Layout, GenericEnv) {
+                    RecursiveTypeProperties::HasArchetype |
+                        RecursiveTypeProperties::HasConcretePack,
```
It’s actually PackExpansionType that leads us into this situation

nate-chandler replied:
Yeah, it's a PackExpansionType whose pattern is a PackArchetypeType:

```
(tuple_type num_elements=1
  (tuple_type_elt
    (pack_expansion_type
      (pattern=pack_archetype_type address=0x127045b10
        (interface_type=dependent_member_type assoc_type=Swift.(file).Sequence.Element
          (base=generic_type_param_type depth=0 index=0 decl=main.(file).tupleExpansionWithMemberType(seqs:elts:).T@main.swift:1:47 pack)) name=(each T).Element
)
      (count=pack_archetype_type address=0x127045880
        (interface_type=generic_type_param_type depth=0 index=0 decl=main.(file).tupleExpansionWithMemberType(seqs:elts:).T@main.swift:1:47 pack) conforms_to=Swift.(file).Sequence name=each T
))))
```

nate-chandler committed:
[AST] NFC: Refactor pack recursive properties.

Eliminated HasConcretePack and added HasPack and HasPackArchetype.
Renamed the old `hasPack` to `hasAnyPack`; as before, it means that the
type has a parameter pack, a pack, or a pack archetype.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

NRC1978 commented:
The beach
